wire-grpc-tests/src/test/java/com/squareup/wire/GrpcBlockingClientTest.kt

```diff
@@ -0,0 +1,284 @@
+    receiveChannel.read()
+    receiveChannel.close()
+    mockService.awaitSuccessBlocking()
+    assertThat(callReference.get()?.isCanceled()).isTrue()
+  }
+
+  @Test
+  fun cancelDuplexBeforeRequestCompletes() {
+    mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RouteChat"))
+
+    val (_, responseChannel) = routeGuideService.RouteChat()
+    // We wait for mockService to process our actions.
+    Thread.sleep(500)
+    responseChannel.close()
+    mockService.awaitSuccessBlocking()
+    assertThat(callReference.get()?.isCanceled()).isTrue()
+  }
+
+  @Test
+  fun cancelDuplexAfterRequestCompletes() {
+    mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RouteChat"))
+    mockService.enqueueReceiveNote(message = "marco")
+    mockService.enqueue(ReceiveComplete)
+
+    val (requestChannel, responseChannel) = routeGuideService.RouteChat()
+    requestChannel.write(RouteNote(message = "marco"))
+    requestChannel.close()
+    // We wait for mockService to process our actions.
+    Thread.sleep(500)
+
+    responseChannel.close()
+    mockService.awaitSuccessBlocking()
+    assertThat(callReference.get()?.isCanceled()).isTrue()
+  }
+
+  @Test
+  fun cancelDuplexBeforeResponseCompletes() {
+    mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RouteChat"))
+    mockService.enqueueReceiveNote(message = "marco")
+    mockService.enqueueSendNote(message = "polo")
+    mockService.enqueueReceiveNote(message = "rené")
+    mockService.enqueue(ReceiveComplete)
+
+    val (requestChannel, responseChannel) = routeGuideService.RouteChat()
+    requestChannel.write(RouteNote(message = "marco"))
+    assertThat(responseChannel.read()).isEqualTo(RouteNote(message = "polo"))
+    requestChannel.write(RouteNote(message = "rené"))
+    requestChannel.close()
+
+    // We wait for mockService to process our actions.
+    Thread.sleep(500)
+
+    responseChannel.close()
+    mockService.awaitSuccessBlocking()
+    assertThat(callReference.get()?.isCanceled()).isTrue()
+  }
+
+  @Test
+  fun cancelDuplexAfterResponseCompletes() {
+    mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RouteChat"))
+    mockService.enqueueReceiveNote(message = "marco")
+    mockService.enqueueSendNote(message = "polo")
+    mockService.enqueueReceiveNote(message = "rené")
+    mockService.enqueueSendNote(message = "lacoste")
+    mockService.enqueue(ReceiveComplete)
+    mockService.enqueue(SendCompleted)
+
+    val (requestChannel, responseChannel) = routeGuideService.RouteChat()
+    requestChannel.write(RouteNote(message = "marco"))
+    assertThat(responseChannel.read()).isEqualTo(RouteNote(message = "polo"))
+    requestChannel.write(RouteNote(message = "rené"))
+    assertThat(responseChannel.read()).isEqualTo(RouteNote(message = "lacoste"))
+    requestChannel.close()
+    assertThat(responseChannel.read()).isNull()
+
+    // We wait for mockService to process our actions.
+    Thread.sleep(500)
```

Comment:
Lots of sleeps :(

Reply:
Yeah. Maybe we want to be able to give the mock server a CountDownLatch or something.